### PR TITLE
NAS-127208 / 24.10 / Refactor VmListComponent

### PR DESCRIPTION
**Pull request by denysbutenko**
Make advanced tests for regressions

### CODE REVIEW

File: src/app/pages/vm/vm-list/vm-list.component.spec.ts [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+import { IxTable2Harness } from 'app/modules/ix-table2/components/ix-table2/ix-table2.harness';
+import { IxTable2Module } from 'app/modules/ix-table2/ix-table2.module';
+import { AppLoaderModule } from 'app/modules/loader/app-loader.module';
+import { VmWizardComponent } from 'app/pages/vm/vm-wizard/vm-wizard.component';
+import { IxSlideInService } from 'app/services/ix-slide-in.service';
+import { SystemGeneralService } from 'app/services/system-general.service';
+import { VmService } from 'app/services/vm.service';
+import { VmListComponent } from './vm-list.component';
+
+const virtualMachines: VirtualMachine[] = [
```
Comment (AlexKarpov98) [inline src/app/pages/vm/vm-list/vm-list.component.spec.ts:23]: please remove unused props from the object.

File: src/app/pages/vm/vm-list/vm-details-row/vm-details-row.component.spec.ts [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+} from 'app/enums/vm.enum';
+import { VirtualMachine } from 'app/interfaces/virtual-machine.interface';
+import { VmNicDevice, VmDisplayDevice, VmDiskDevice } from 'app/interfaces/vm-device.interface';
+import { VmEditFormComponent } from 'app/pages/vm/vm-edit-form/vm-edit-form.component';
+import { DeleteVmDialogComponent } from 'app/pages/vm/vm-list/delete-vm-dialog/delete-vm-dialog.component';
+import { IxSlideInService } from 'app/services/ix-slide-in.service';
+import { VmService } from 'app/services/vm.service';
+import { VirtualMachineDetailsRowComponent } from './vm-details-row.component';
+
+const virtualMachine = {
```
Comment (AlexKarpov98) [inline src/app/pages/vm/vm-list/vm-details-row/vm-details-row.component.spec.ts:21]: make it smaller, removed unused props from the object.

**Review by AlexKarpov98 — CHANGES_REQUESTED** [review]
@denysbutenko - good job, please fix issues I've found.

1. Page title is missing when there is no data available:
Screenshot 2024-02-14 at 14 56 09

2. When State action failed, toggler is not returned to previous state.

https://github.com/truenas/webui/assets/22980553/51448603-258d-41d5-9c11-4b57f885f23e



3. I change `Start on Boot` toggler and nothing changes, if I reload the page - it stays at default values.
Screenshot 2024-02-14 at 15 02 50

4. No checks for Readonly user.
Screenshot 2024-02-14 at 15 14 30

5. Sorting is not always working, see video:

https://github.com/truenas/webui/assets/22980553/8a3b5701-61b9-4c1a-80c7-b9f243ad5cc6



6. State toggler is not updated once I `Start`ed VM, see:

https://github.com/truenas/webui/assets/22980553/a9570ef5-18a0-492c-942c-faa3967282b5

7. Search is working not fully as expected, I find VM which has nothing related to the search at the first glance.

https://github.com/truenas/webui/assets/22980553/693212c1-4b37-4dc5-a64a-d5a97e5eb2db

8. On mobile (details row) buttons are not looking good and overlapping main content.
Screenshot 2024-02-14 at 15 23 08
Screenshot 2024-02-14 at 15 23 49

**Comment by denysbutenko** [comment]
@AlexKarpov98 Most of the regressions are fixed. Basic role support was added. For advanced cases – I'll create a ticket to revisit roles for VM forms and dialogs.

**Comment by AlexKarpov98** [comment]
@denysbutenko - you can remove sorting from `Display Port` as well.

https://github.com/truenas/webui/assets/22980553/62684322-f6bb-4ea1-a62e-f76e8a4c5692

**Comment by AlexKarpov98** [comment]
@denysbutenko  Screenshot 2024-02-15 at 14 36 21
For `Running` and `Start on Boot` - please add `ixRequiresRoles` as well.

**Review by AlexKarpov98 — APPROVED** [review]
Looks good, thanks 👍

**Comment by denysbutenko** [comment]
Linked tickets:

- [NAS-127372 Add new endpoint to download VM logs](https://ixsystems.atlassian.net/browse/NAS-127373)
- [NAS-127373 WebUI: Add role support for VM components](https://ixsystems.atlassian.net/browse/NAS-127373)

**Pull request merged**